### ISSUE TITLE
Cannot run on Windows 10 and Java 8

### ISSUE THREAD

**Issue by Ull** [sarahtattersall/PIPE#97]
On Windows 10, java 1.8.0_65, if I try to execute `java -jar PIPE-gui-5.0.0.jar` I get an exception.

```
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException
        at javax.swing.ImageIcon.<init>(Unknown Source)
        at pipe.views.PipeApplicationView.<init>(PipeApplicationView.java:129)
        at pipe.views.PipeApplicationBuilder.build(PipeApplicationBuilder.java:53)
        at Pipe.<init>(Pipe.java:17)
        at Pipe.<init>(Pipe.java:9)
        at Pipe$1.run(Pipe.java:31)
        at java.awt.event.InvocationEvent.dispatch(Unknown Source)
        at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
        at java.awt.EventQueue.access$500(Unknown Source)
        at java.awt.EventQueue$3.run(Unknown Source)
        at java.awt.EventQueue$3.run(Unknown Source)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
        at java.awt.EventQueue.dispatchEvent(Unknown Source)
        at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
        at java.awt.EventDispatchThread.run(Unknown Source)
```

Tried with other version of Pipe, didn't work.

**Comment by sjdayday** [comment]
PIPE has some known problems with Java 8 -- you might try with Java 7.
Thanks for the reporting the problem; I'll try to get it fixed in the next
two weeks.  Steve

On Thu, Oct 29, 2015 at 10:26 AM, Ull <notifications@github.com> wrote:

> On Windows 10, java 1.8.0_65, if I try to execute java -jar
> PIPE-gui-5.0.0.jar I get an exception.
>
> Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException
>         at javax.swing.ImageIcon.<init>(Unknown Source)
>         at pipe.views.PipeApplicationView.<init>(PipeApplicationView.java:129)
>         at pipe.views.PipeApplicationBuilder.build(PipeApplicationBuilder.java:53)
>         at Pipe.<init>(Pipe.java:17)
>         at Pipe.<init>(Pipe.java:9)
>         at Pipe$1.run(Pipe.java:31)
>         at java.awt.event.InvocationEvent.dispatch(Unknown Source)
>         at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
>         at java.awt.EventQueue.access$500(Unknown Source)
>         at java.awt.EventQueue$3.run(Unknown Source)
>         at java.awt.EventQueue$3.run(Unknown Source)
>         at java.security.AccessController.doPrivileged(Native Method)
>         at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
>         at java.awt.EventQueue.dispatchEvent(Unknown Source)
>         at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
>         at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
>         at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
>         at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
>         at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
>         at java.awt.EventDispatchThread.run(Unknown Source)
>
> Tried with other version of Pipe, didn't work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sarahtattersall/PIPE/issues/97>.
>



-- 
Steve Doubleday
UC Irvine
Mathematical Behavioral Sciences
stevedoubleday@gmail.com
818-648-8381

**Comment by Ull** [comment]
I tried Java 1.7 32 & 64 bits, always the same exceptions
```
C:\Users\thiba\Desktop {git}
{lamb} java -jar PIPE-gui-5.0.0.jar
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException
        at javax.swing.ImageIcon.<init>(Unknown Source)
        at pipe.views.PipeApplicationView.<init>(PipeApplicationView.java:129)
        at pipe.views.PipeApplicationBuilder.build(PipeApplicationBuilder.java:53)
        at Pipe.<init>(Pipe.java:17)
        at Pipe.<init>(Pipe.java:9)
        at Pipe$1.run(Pipe.java:31)
        at java.awt.event.InvocationEvent.dispatch(Unknown Source)
        at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
        at java.awt.EventQueue.access$300(Unknown Source)
        at java.awt.EventQueue$3.run(Unknown Source)
        at java.awt.EventQueue$3.run(Unknown Source)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.security.ProtectionDomain$1.doIntersectionPrivilege(Unknown Source)
        at java.awt.EventQueue.dispatchEvent(Unknown Source)
        at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
        at java.awt.EventDispatchThread.run(Unknown Source)

C:\Users\thiba\Desktop {git}
{lamb} java -version
java version "1.7.0_79"
Java(TM) SE Runtime Environment (build 1.7.0_79-b15)
Java HotSpot(TM) Client VM (build 24.79-b02, mixed mode, sharing)
```

**Comment by dcduba** [comment]
In PipeApplicationView the icon is set as:

    this.setIconImage(new ImageIcon(getImageURL("icon.png").getImage());

This is getImageURL:

    private URL getImageURL(String name) {
        return this.getClass().getResource(PIPEConstants.IMAGE_PATH + name);
    }

Which will return null. `PIPEConstants.IMAGE_PATH == "\images\"`. The file is located in `pipe-gui\src\main\resources\images`, but I am *guessing* that in Windows 10, it is not searching in the right place.

**Comment by sjdayday** [comment]
Thank you for reporting this.  What version of java are you running?

best,
Steve

On Thu, Dec 10, 2015 at 2:39 PM, dcduba <notifications@github.com> wrote:

> In PipeApplicationView the icon is set as:
>
> this.setIconImage(new ImageIcon(getImageURL("icon.png").getImage());
>
> This is getImageURL:
>
> private URL getImageURL(String name) {
>     return this.getClass().getResource(PIPEConstants.IMAGE_PATH + name);
> }
>
> Which will return null. PIPEConstants.IMAGE_PATH == "\images\". The file
> is located in pipe-gui\src\main\resources\images, but I am *guessing*
> that in Windows 10, it is not searching in the right place.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sarahtattersall/PIPE/issues/97#issuecomment-163774599>
> .
>



-- 
Steve Doubleday
UC Irvine
Mathematical Behavioral Sciences
stevedoubleday@gmail.com
818-648-8381

**Comment by sjdayday** [comment]
That was dumb -- I just read the full title of your email :)

Others have reported similar problems; I haven't tested this against either
a current level of windows or java 8.  I hope to have time to research this
soon.

Thanks.

Steve

On Thu, Dec 10, 2015 at 3:53 PM, Steve Doubleday <stevedoubleday@gmail.com>
wrote:

> Thank you for reporting this.  What version of java are you running?
>
> best,
> Steve
>
> On Thu, Dec 10, 2015 at 2:39 PM, dcduba <notifications@github.com> wrote:
>
>> In PipeApplicationView the icon is set as:
>>
>> this.setIconImage(new ImageIcon(getImageURL("icon.png").getImage());
>>
>> This is getImageURL:
>>
>> private URL getImageURL(String name) {
>>     return this.getClass().getResource(PIPEConstants.IMAGE_PATH + name);
>> }
>>
>> Which will return null. PIPEConstants.IMAGE_PATH == "\images\". The file
>> is located in pipe-gui\src\main\resources\images, but I am *guessing*
>> that in Windows 10, it is not searching in the right place.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/sarahtattersall/PIPE/issues/97#issuecomment-163774599>
>> .
>>
>
>
>
> --
> Steve Doubleday
> UC Irvine
> Mathematical Behavioral Sciences
> stevedoubleday@gmail.com
> 818-648-8381
>



-- 
Steve Doubleday
UC Irvine
Mathematical Behavioral Sciences
stevedoubleday@gmail.com
818-648-8381

**Comment by dcduba** [comment]
The constant used uses `Path.separator`, which is `\` on windows instead of `/` on linux. `getResource(..)` expects a `/` as first character if it is an absolute path, even if the seperator is different. This results in it prepending the class location (`/main/view`) to the url, which results in the funky looking absolute path `/path/to/main/view/\\images\\icon.png`. Needless to say that will not work. Changing the first separator in the constants to a literal `/` will fix the problem.

I'll try to add a pull request later today for both this and java 8, as the problems with java 8 seem to be literally 2 incomplete types for `argThat<...>` and a bunch of javadoc statements.

**Comment by dcduba** [comment]
I see that there is already a pull request for this issue. Related code that is called:

    public java.net.URL getResource(String name) {
        name = resolveName(name);
        ClassLoader cl = getClassLoader0();
        if (cl==null) {
            // A system class.
            return ClassLoader.getSystemResource(name);
        }
        return cl.getResource(name);
    }

which calls:

    private String resolveName(String name) {
        if (name == null) {
            return name;
        }
        if (!name.startsWith("/")) {
            Class<?> c = this;
            while (c.isArray()) {
                c = c.getComponentType();
            }
            String baseName = c.getName();
            int index = baseName.lastIndexOf('.');
            if (index != -1) {
                name = baseName.substring(0, index).replace('.', '/')
                    +"/"+name;
            }
        } else {
            name = name.substring(1);
        }
        return name;
    }

I don't think it makes sense to do it in any other way, as the `/` is hardcoded in the code that is called. Are these constants ever used outside a `getResource(..)` call?

**Comment by sjdayday** [comment]
Thank you for chasing this down so thoroughly.   

I had been assuming this was a Java 8 problem, but now I'm guessing PIPE 5 has never successfully run on Windows, because of this error.  I re-created the error under Windows, and verified that changing from using File.separator to a hard-coded "/" fixes the problem.  The Class.getResource method doc is consistent with your interpretation: 
https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#getResource-java.lang.String-

Currently all access to both paths in PIPEConstants is through the getResource(..) call, so the simplest thing is probably to replace all the file separators (not just the leading one) with "/".  I'll put in the fix, but won't re-release the uber jar until the other Java 8 issues are addressed. 

Steve

**Comment by OnnOElzinga** [comment]
Dear Sarah and Steve, would you be able to give some hints/patch or a script to get PIPE running on win/7 64? 
Thanks a lot in advance, cheers, Onno.

**Comment by sjdayday** [comment]
What error do you get?

**Comment by OnnOElzinga** [comment]
Hi :+1: 

C:\Users\Onno\Documents\prive\leiden\statsci\PIPE>java -jar PIPE-gui-5.0.0-beta-
3.jar
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException
        at javax.swing.ImageIcon.<init>(Unknown Source)
        at pipe.views.PipeApplicationView.<init>(PipeApplicationView.java:173)
        at Pipe.<init>(Pipe.java:27)
        at Pipe.<init>(Pipe.java:8)
        at Pipe$1.run(Pipe.java:42)
        at java.awt.event.InvocationEvent.dispatch(Unknown Source)
        at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
        at java.awt.EventQueue.access$500(Unknown Source)
        at java.awt.EventQueue$3.run(Unknown Source)
        at java.awt.EventQueue$3.run(Unknown Source)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionP
rivilege(Unknown Source)
        at java.awt.EventQueue.dispatchEvent(Unknown Source)
        at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
        at java.awt.EventDispatchThread.run(Unknown Source)

**Comment by sjdayday** [comment]
Yes, that's the same problem.  I'm hoping to get that fixed this week.  Steve

**Comment by OnnOElzinga** [comment]
Thanks, that would mean I can spend my summer vacation responsibly ;-)

**Comment by sjdayday** [comment]
This is fixed in [PIPE release 5.0.2](https://github.com/sarahtattersall/PIPE/releases/tag/PIPE-5.0.2).  I've been able to verify it under Windows 7 and Java 8.  Thanks very much to [dcduba](https://github.com/dcduba) for all the research on this.  Sorry it took me so long to get to it.
